src/components/DataRoom/IndustryComparison/DotGrid.tsx

```diff
@@ -0,0 +1,107 @@
+import type { MotionValue } from 'framer-motion'
+import { useEffect, useMemo, useRef, useState } from 'react'
+import { createDots } from './utils/createDots'
+import { updateCanvasDimensions } from './utils/updateCanvasDimensions'
+import { drawDots } from './utils/drawDots'
+import { lerp } from './utils/lerp'
+
+export default function DotGrid({
+  containerRef,
+  isMobile,
+  scrollYProgress,
+}: {
+  containerRef: React.RefObject<HTMLDivElement>
+  isMobile: boolean
+  scrollYProgress?: MotionValue<number>
+}) {
+  const canvasRef = useRef<HTMLCanvasElement>(null)
+  const [dimensions, setDimensions] = useState({ width: 0, height: 0 })
+  const [mousePosition, setMousePosition] = useState({ x: 0, y: 0 })
+  const [lerpedMousePosition, setLerpedMousePosition] = useState({ x: 0, y: 0 })
+
+  const dots = useMemo(() => createDots(dimensions, isMobile), [dimensions, isMobile])
+
+  useEffect(() => {
+    const updateDimensions = () => {
+      const newDimensions = updateCanvasDimensions(canvasRef, containerRef)
+      setDimensions(newDimensions)
+    }
+
+    const handleMouseMove = (e: MouseEvent) => {
+      if (canvasRef.current && containerRef.current) {
+        const rect = containerRef.current.getBoundingClientRect()
+        setMousePosition({ x: e.clientX - rect.left, y: e.clientY - rect.top })
+      }
+    }
+
+    const updateMobileMousePosition = () => {
+      if (scrollYProgress && isMobile && containerRef.current) {
```

Comment:
We could use early returns here to reduce indentation, e.g.
```suggestion
      if (canvasRef.current && containerRef.current) return
      const rect = containerRef.current.getBoundingClientRect()
      setMousePosition({ x: e.clientX - rect.left, y: e.clientY - rect.top })
    }

    const updateMobileMousePosition = () => {
      if (scrollYProgress && isMobile && containerRef.current) return
```